### PR TITLE
Variable path searcher

### PR DESCRIPTION
**Pull request by ddn0**
This implementation of variable path matching is implemented as an
extension of the existing path matcher for simplicity.

It avoids path materialization but depending on the matching order it
may re-evaluate a path count search multiple times for the same source.
This can be alleviated by a different matching order or by a future
optimization where variable path matches are done separately and
embeddings from different subqueries are joined. The latter optimization
will require more a distributed hash join or equivalent which is not yet
supported in libgluon.

### CODE REVIEW

File: libquery/src/SubgraphQuery.cpp [hunk 1]
```diff
@@ -454,71 +430,359 @@ class SubgraphQueryMiner {
...
+      // When source == sink is a valid path, we still need to check if other
+      // edge constraints are satisfied.
+      if (data_node == source) {
+        if (!CanAdd(sink_order, emb, data_node, neighbor_index, matched)) {
+          return;
+        }
+      }
+
+      auto& local_list = *out_list.getLocal();
+      for (uint32_t i = 0; i < count; ++i) {
```
Comment (roshandathathri) [inline libquery/src/SubgraphQuery.cpp:644]: Don't think this is the semantics. As far as I understand, sink is added only once. The count is a property of that matched node which can be subsequently used.

File: libquery/src/SubgraphQuery.cpp [hunk 2]
```diff
@@ -454,71 +430,359 @@ class SubgraphQueryMiner {
...
+    MatchedNeighbors matched = NextNeighbors(num_matched);
+
+    if (matched.neighbors.empty()) {
+      // Disconnected query node
+      return AddNode(GetQueryVertex(num_matched), in_list);
+    }
+
+    in_list = ExtendVariablePaths(num_matched, matched, std::move(in_list));
+
+    in_list = ExtendUnitPaths(num_matched, matched, std::move(in_list));
```
Comment (roshandathathri) [inline libquery/src/SubgraphQuery.cpp:668]: Why is this required when ExtendVariablePaths already uses CanAdd() to check for the unit paths?

**Review by roshandathathri — APPROVED** [review]
LGTM except a couple of questions in variable length path functions. The rest of the comments are nits.

File: libquery/src/SubgraphQuery.cpp (hunk 1 above)
Comment (ddn0) [reply libquery/src/SubgraphQuery.cpp:644]: I got the duplicate sink behavior from observing Neo4j. In the below there are two paths from "Otto" to "Alim", and in the resulting tables, this row is repeated for each path:

```
neo4j> MATCH (person:Person {firstName:"Otto"})-[path:KNOWS*1..2]->(friend:Person) RETURN path,friend.firstName;
+---------------------------------------------------------------------------------------------------------------------------+
| path                                                                                                   | friend.firstName |
+---------------------------------------------------------------------------------------------------------------------------+
| [[:KNOWS {creationDate: 2012-03-23T01:18:40.065Z}]]                                                    | "Joakim"         |
| [[:KNOWS {creationDate: 2012-03-23T01:18:40.065Z}], [:KNOWS {creationDate: 2012-04-26T21:24:21.722Z}]] | "Alim"           |
| [[:KNOWS {creationDate: 2011-06-22T09:25:09.581Z}]]                                                    | "Bryn"           |
| [[:KNOWS {creationDate: 2011-06-22T09:25:09.581Z}], [:KNOWS {creationDate: 2011-11-07T11:10:46.510Z}]] | "Alim"           |
| [[:KNOWS {creationDate: 2011-06-22T09:25:09.581Z}], [:KNOWS {creationDate: 2012-08-28T02:35:08.233Z}]] | "Almira"         |
| [[:KNOWS {creationDate: 2011-06-22T09:25:09.581Z}], [:KNOWS {creationDate: 2012-03-29T16:36:25.431Z}]] | "Hans"           |
| [[:KNOWS {creationDate: 2011-06-22T09:25:09.581Z}], [:KNOWS {creationDate: 2012-05-03T14:44:49.184Z}]] | "Ken"            |
+---------------------------------------------------------------------------------------------------------------------------+

7 rows available after 32 ms, consumed after another 3 ms

neo4j> MATCH (person:Person {firstName:"Otto"})-[path:KNOWS*1..2]->(friend:Person) RETURN friend.firstName;
+------------------+
| friend.firstName |
+------------------+
| "Joakim"         |
| "Alim"           |
| "Bryn"           |
| "Alim"           |
| "Almira"         |
| "Hans"           |
| "Ken"            |
+------------------+

7 rows available after 26 ms, consumed after another 2 ms

```

File: libquery/src/SubgraphQuery.cpp (hunk 2 above)
Comment (ddn0) [reply libquery/src/SubgraphQuery.cpp:668]: This actually didn't occur to me when I was writing `ExtendVariablePaths` how'd much they'd overlap. In some sense, `ExtendUnitPaths` and `ExtendVariable` paths represent two different parallelization strategies that share the same checking function (CanAdd).

- Are you asking that we should get rid of one for the other? I would guess we want to keep both because they are specialized for different workloads.

- Are you asking why we would call `ExtendUnitPaths` if `ExtendVariablePaths` has already extended a node? If so, the check to skip work in `ExtendUnitPaths` if `ExtendVaraiblePaths` did something is in side `ExtendUnitPaths`. I can pull this bit out to make the control flow more obvious. Originally I had the idea that you might want to play around with the order of these two functions, e.g., for a node that has both incoming unit and variable paths match unit paths and then filter for variable paths or vice versa, but the current ordering seems better in general, so we can just stick with that.

File: libquery/src/SubgraphQuery.cpp [hunk 3]
```diff
@@ -454,71 +430,359 @@ class SubgraphQueryMiner {
...
+   */
+  void ExtendVariablePathSource(const QueryEmbedding& emb, VertexId source,
+                                VertexId sink_qnode,
+                                const QueryEdgeData& query_edge,
+                                unsigned neighbor_index,
+                                const MatchedNeighbors& matched,
+                                ConcurrentQueryEmbeddingList& out_list) {
+
+    galois::InsertBag<uint32_t> sinks;
+    galois::LargeArray<bool> is_sink;
```
[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/src/SubgraphQuery.cpp:521]: Back in my day.....

File: libquery/include/galois/graphs/QueryGraph.h [hunk 4]
```diff
@@ -85,10 +82,38 @@ struct QueryEdgeData {
   //! Direction of the edge
   QueryEdgeDirection edge_direction{galois::DIRECTED_EDGE};
 
+  std::optional<unsigned> var_length_start;
+  std::optional<unsigned> var_length_end;
```
[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/include/galois/graphs/QueryGraph.h:86]: Without the optional? E.g., represent unbounded end length as MAX_INT?

That's fine with me, but I'm just confirming.

[the remark this answers is not in the document]
Comment (roshandathathri) [reply libquery/include/galois/graphs/QueryGraph.h:86]: Yes but that's just my opinion.

File: libquery/src/SubgraphQuery.cpp (hunk 1 above)
Comment (roshandathathri) [reply libquery/src/SubgraphQuery.cpp:644]: Thanks!

File: libquery/src/SubgraphQuery.cpp (hunk 2 above)
Comment (roshandathathri) [reply libquery/src/SubgraphQuery.cpp:668]: I meant the last thing you said. Yes, it would be better if the control flow is more explicit here. 

Yes, we would definitely want to keep both.

File: libquery/src/SubgraphQuery.cpp [hunk 5]
```diff
@@ -454,71 +445,363 @@ class SubgraphQueryMiner {
...
+    unsigned num_matched = in_list.GetEmbeddingSize();
+
+    MatchedNeighbors matched = NextNeighbors(num_matched);
+
+    if (matched.neighbors.empty()) {
+      // Disconnected query node
+      return AddNode(GetQueryVertex(num_matched), in_list);
+    }
+
+    if (num_matched > 0 &&
```
Comment (roshandathathri) [inline libquery/src/SubgraphQuery.cpp:673]: Isn't this always true?
